platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/snapshotter/MapSnapshotter.java

```diff
@@ -269,43 +279,114 @@ public void cancel() {
    * @param mapSnapshot the map snapshot to draw the overlay on
    */
   protected void addOverlay(MapSnapshot mapSnapshot) {
-    Bitmap original = mapSnapshot.getBitmap();
-    Canvas canvas = new Canvas(original);
-    addLogo(canvas, original);
+    Bitmap snapshot = mapSnapshot.getBitmap();
+    Canvas canvas = new Canvas(snapshot);
+    DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
+    int margin = (int) displayMetrics.density * LOGO_MARGIN_DP;
+
+    Logo logo = createScaledLogo(snapshot);
+    TextView longText = createTextView(mapSnapshot, false, logo.getScale());
+    TextView shortText = createTextView(mapSnapshot, true, logo.getScale());
+
+    AttributionMeasure measure = new AttributionMeasure.Builder()
+      .setSnapshot(snapshot)
+      .setLogo(logo.getLarge())
+      .setLogoSmall(logo.getSmall())
+      .setTextView(longText)
+      .setTextViewShort(shortText)
+      .setMarginPadding(margin)
+      .build();
+
+    AttributionLayout layout = measure.measure();
+
+    // draw logo
```

Comment:
`addOverlay` is getting long.
What about extracting these blocks of code into private methods and give them a name based on the comments? This way `addOverlay` will be easier to read and understand and comments will become superfluous so they won't be necessary.